lib/class-wp-rest-url-details-controller.php

```diff
@@ -0,0 +1,291 @@
+					'readonly'    => true,
+				),
+			),
+		);
+
+		$this->schema = $schema;
+
+		return $this->add_additional_fields_schema( $this->schema );
+	}
+
+	/**
+	 * Retrieves the contents of the <title> tag from the HTML
+	 * response.
+	 *
+	 * @param WP_REST_REQUEST $request Full details about the request.
+	 * @return String|WP_Error The title text or an error.
+	 */
+	public function parse_url_details( $request ) {
+
+		$url = untrailingslashit( $request->get_param( 'url' ) );
+
+		if ( empty( $url ) ) {
+			return new WP_Error( 'rest_invalid_url', __( 'Invalid URL', 'gutenberg' ), array( 'status' => 404 ) );
+		}
+
+		// Transient per URL.
+		$cache_key = $this->build_cache_key_for_url( $url );
+
+		// Attempt to retrieve cached response.
+		$cached_response = $this->get_cache( $cache_key );
+
+		if ( ! empty( $cached_response ) ) {
+			$from_cache = true;
+			$response = rest_ensure_response( json_decode( $cached_response, true ) );
+
+			/**
+			 * Filters the URL data for the response.
+			 *
+			 * @param WP_REST_Response $response The response object.
+			 * @param string           $url      The requested URL.
+			 * @param bool             $from_cache Whether the response is from the cache.
+			 * @param WP_REST_Request  $request  Request object.
+			 */
+			return apply_filters( 'rest_url_details_prepare_response', $response, $url, $from_cache, $request );
+		}
+
+		// If we're this far the response is not from the cache.
+		$from_cache = false;
+
+		$response_body = $this->get_remote_url( $url );
+
+		// Exit if we don't have a valid body or it's empty.
+		if ( is_wp_error( $response_body ) || empty( $response_body ) ) {
+			return $response_body;
+		}
+
+		/**
+		 * Filters the data returned for a given URL.
+		 *
+		 * Can be used to modify the data retrieved for a given URL.
+		 * For example, you may wish to retrieve additional information
+		 * from the response body at a given URL over and above the `<title>`.
+		 *
+		 * @param array $data The default data retrieved for the given URL.
+		 * @param array $response_body The response body (HTML) for the given URL.
+		 * @param string $url The attempted URL.
+		 */
+		$data = apply_filters(
+			'rest_url_details_url_data',
+			array(
+				'title' => $this->get_title( $response_body ),
+			),
+			$response_body,
+			$url
+		);
+
+		$data = $this->add_additional_fields_to_object( $data, $request );
+
+		// Cache the valid response.
+		$this->set_cache( $cache_key, $data );
+
+		// Wrap the data in a response object.
+		$response = rest_ensure_response( $data );
+
+		/**
+		 * Filters the URL data for the response.
+		 *
+		 * @param WP_REST_Response $response The response object.
+		 * @param string           $url      The requested URL.
+		 * @param bool             $from_cache Whether the response is from the cache.
+		 * @param WP_REST_Request  $request  Request object.
+		 */
+		return apply_filters( 'rest_url_details_prepare_response', $response, $url, $from_cache, $request );
```

Comment:
```suggestion
		return apply_filters( 'rest_prepare_url_details', $response, $url, $from_cache, $request );
```

Also, is the `from_cache` parameter necessary? I don't think that should be something plugins rely on.

Reply:
I thought it might be useful to have the ability to distinguish between response from cache or not. No other reason. Would that _not_ be useful?

Reply:
I think the caching should be considered an implementation detail.

Reply:
TBH we don't have any use case for needing the `from_cache` identifier so let's reduce the complexity and not overengineer. I'll remove it.

Reply:
Removed `$from_cache` and updated filter.

Reply:
Done

Reply:
I still see `$from_cache` here